node/silkworm/stagedsync/stage_interhashes/trie_cursor.cpp

```diff
@@ -0,0 +1,336 @@
+        return false;
+    }
+
+    ByteView db_cursor_key{db::from_slice(data.key)};  // Save db_cursor_ key ...
+    db_cursor_key.remove_prefix(prefix_.length());     // ... and remove prefix_ so we have node key
+    if (seek_key.empty() && !db_cursor_key.empty()) {
+        // Note ! an empty seek_key means we're looking for a root node with empty key which does not exist
+        return false;
+    }
+
+    ByteView db_cursor_val{db::from_slice(data.value)};  // Save db_cursor_ value
+    level_ += seek_key.empty() ? 0 : 1u;                 // Down one level for child node. Stay at zero for root node
+    auto& new_node{sub_nodes_[level_]};
+    new_node.parse(db_cursor_key, db_cursor_val);
+    return true;
+}
+
+void TrieCursor::db_delete(SubNode& node) {
+    if (!node.deleted && collector_) {
+        buffer_.assign(prefix_).append(node.key);
+        collector_->collect({buffer_, Bytes{}});
+        node.deleted = true;
+    }
+}
+
+bool TrieCursor::consume(SubNode& node) {
+    if (node.has_hash()) {
+        buffer_.assign(prefix_).append(node.full_key());
+        auto [has_changes, next_created]{changed_list_->contains_and_next_marked(buffer_)};
+        if (!has_changes) {
+            skip_state_ = skip_state_ && key_is_before(buffer_, next_created_);
+            std::swap(next_created_, next_created);
+            return true;
+        }
+    }
+    db_delete(node);
+    return false;
+}
+
+bool TrieCursor::key_is_before(ByteView k1, ByteView k2) {
+    if (k1.is_null()) {
+        return false;
+    }
+    if (k2.is_null()) {
+        return true;
+    }
+    return k1 < k2;
+}
+std::optional<Bytes> TrieCursor::increment_nibbled_key(const ByteView origin) {
+    Bytes ret{};
+    auto rit{std::find_if(origin.rbegin(), origin.rend(), [](uint8_t nibble) { return nibble != 0xf; })};
+    if (rit == origin.rend()) {
+        // Overflow
+        return std::nullopt;
+    }
+    auto count{static_cast<size_t>(std::distance(origin.begin(), rit.base()))};
+    ret.assign(origin.substr(0, count));
+    ++ret.back();
+    return ret;
+}
+std::optional<Bytes> TrieCursor::first_uncovered() {
+    if (skip_state_) {
+        return std::nullopt;
+    }
+
+    // This is intended. Don't want an empty origin to be marked as overflown
+    if (prev_key_.empty()) {
+        return prev_key_;
+    }
+
+    const auto incremented_nibbled_key{increment_nibbled_key(prev_key_)};
+    if (incremented_nibbled_key.has_value()) {
+        return pack_nibbles(incremented_nibbled_key.value());
+    }
+    return incremented_nibbled_key;
+}
+
+void TrieCursor::SubNode::reset() {
+    key = ByteView();
+    value = ByteView();
+    root_hash = ByteView();
+    hashes = ByteView();
+    state_mask = 0;
+    tree_mask = 0;
+    hash_mask = 0;
+    child_id = -1;
+    max_child_id = 0x10;  // Traverse all node for ephemeral ones.
+    hash_id = -1;
+    deleted = false;
+}
+
+void TrieCursor::SubNode::parse(ByteView k, ByteView v) {
```

Comment:
Please reduce code duplication between `TrieCursor::SubNode::parse` and `Node::from_encoded_storage`. Perhaps `SubNode` should be a subclass of `Node`.

Reply:
Actually SubNode of TrieCursor and trie::Node are totally different beasts.
The first has no allocations and all data is held in ByteViews.
The second copies hashes in a vector every-time

Reply:
Besides trie::Node (as being also part of WASM build) does not implement exceptions so it makes hard to understand why a  parse from encoded storage has returned std::nullopt.

A solution would be to return a parsing code (pretty much like we do with rlp::decode) for parsing trie::Node
(though it does not remove the copying of hashes in a vector).

Reply:
Hmm. IMHO it'd still be great to remove the code duplication unless that complicates things even more. What if we 1) return error codes in `Node::from_encoded_storage` and 2) change it to accept a `Node&` as an input/output parameter instead of returning a new instance (similar to what we do in RLP decoding)? With 2) we can probably still use `Node` as a superclass of `SubNode` because we can pre-allocate instances and re-use them in decoding w/o suffering too much memory reallocations.

Reply:
I think it'd be very nice, but it's not critical. Let's try to do it in a follow-up PR.